docs build warnings (treated as an error)

I'm seeing a bunch of warnings when doing a clean docs build (related to types containing a pipe `|`?). These are treated as errors. Everything seems to work fine though, and calling `make` again gives a "green" build. I tried installing the requirements to match the CI, but this didn't seem to fix the problem.

```
t490:msprime $ pwd
/home/grg/src/msprime
t490:msprime $ make -C docs clean
make: Entering directory '/home/grg/src/msprime/docs'
rm -fR _build
make: Leaving directory '/home/grg/src/msprime/docs'
t490:msprime $ make -C docs
make: Entering directory '/home/grg/src/msprime/docs'
PYTHONPATH=/home/grg/src/msprime/.. ./build.sh
Running Jupyter-Book v0.11.2
Source Folder: /home/grg/src/msprime/docs
Config Path: /home/grg/src/msprime/docs/_config.yml
Output Path: /home/grg/src/msprime/docs/_build/html
Running Sphinx v3.3.1
making output directory... done
[etoc] Changing master_doc to 'intro'
[etoc] Excluded 1 extra file(s) not in toc
myst v0.13.7: MdParserConfig(renderer='sphinx', commonmark_only=False, dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, update_mathjax=True, enable_extensions=['colon_fence', 'deflist'], disable_syntax=[], url_schemes=['mailto', 'http', 'https'], heading_anchors=None, html_meta=[], footnote_transition=True, substitutions=[], sub_delimiters=['{', '}'])
loading intersphinx inventory from https://docs.python.org/3/objects.inv...
loading intersphinx inventory from https://tskit.dev/tskit/docs/stable/objects.inv...
loading intersphinx inventory from https://tskit.dev/tutorials/objects.inv...
loading intersphinx inventory from https://stdpopsim.readthedocs.io/en/stable/objects.inv...
loading intersphinx inventory from https://tskit.dev/pyslim/docs/latest/objects.inv...
loading intersphinx inventory from https://numpy.org/doc/stable/objects.inv...
loading intersphinx inventory from https://popsim-consortium.github.io/demes-docs/main/objects.inv...
loading intersphinx inventory from https://popsim-consortium.github.io/demes-spec-docs/main/objects.inv...
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 17 source files that are out of date
updating environment: executing outdated notebooks... Executing: /home/grg/src/msprime/docs/demography.md
Execution Succeeded: /home/grg/src/msprime/docs/demography.md
Executing: /home/grg/src/msprime/docs/logging.md
Execution Succeeded: /home/grg/src/msprime/docs/logging.md
Executing: /home/grg/src/msprime/docs/mutations.md
Execution Succeeded: /home/grg/src/msprime/docs/mutations.md
Executing: /home/grg/src/msprime/docs/likelihoods.md
Execution Succeeded: /home/grg/src/msprime/docs/likelihoods.md
Executing: /home/grg/src/msprime/docs/legacy.md
Execution Succeeded: /home/grg/src/msprime/docs/legacy.md
Executing: /home/grg/src/msprime/docs/ancestry.md
Execution Succeeded: /home/grg/src/msprime/docs/ancestry.md
Executing: /home/grg/src/msprime/docs/rate_maps.md
Execution Succeeded: /home/grg/src/msprime/docs/rate_maps.md
Executing: /home/grg/src/msprime/docs/quickstart.md
Execution Succeeded: /home/grg/src/msprime/docs/quickstart.md
Executing: /home/grg/src/msprime/docs/replication.md
Execution Succeeded: /home/grg/src/msprime/docs/replication.md
done
[new config] 17 added, 0 changed, 0 removed
reading sources... [100%] switch_from_other_simulators
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] switch_from_other_simulators
/home/grg/src/msprime/msprime/ancestry.py:docstring of msprime.SampleSet.population:: WARNING: py:class reference target not found: int | str | None
/home/grg/src/msprime/msprime/ancestry.py:docstring of msprime.SampleSet.time:: WARNING: py:class reference target not found: float | None
/home/grg/src/msprime/msprime/ancestry.py:docstring of msprime.SampleSet.ploidy:: WARNING: py:class reference target not found: int | None
/home/grg/src/msprime/msprime/ancestry.py:docstring of msprime.AncestryModel.duration:: WARNING: py:class reference target not found: float | None
/home/grg/src/msprime/msprime/demography.py:docstring of msprime.Population.name:: WARNING: py:class reference target not found: str | None
/home/grg/src/msprime/msprime/demography.py:docstring of msprime.Population.default_sampling_time:: WARNING: py:class reference target not found: float | None
/home/grg/src/msprime/msprime/demography.py:docstring of msprime.Population.initially_active:: WARNING: py:class reference target not found: bool | None
/home/grg/src/msprime/msprime/demography.py:docstring of msprime.Population.id:: WARNING: py:class reference target not found: int | None
generating indices... genindex done
highlighting module code... [100%] msprime.mutations
writing additional pages... search done
copying images... [100%] _build/jupyter_execute/replication_9_1.svg
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build finished with problems, 8 warnings.
[etoc] missing index.html written as redirect to 'intro.html'

===============================================================================

Building your book, returns a non-zero exit code (1). Look above for the cause.

===============================================================================

make: *** [Makefile:8: dev] Error 1
make: Leaving directory '/home/grg/src/msprime/docs'
```

Calling make again...

```
t490:msprime $ make -C docs
make: Entering directory '/home/grg/src/msprime/docs'
PYTHONPATH=/home/grg/src/msprime/.. ./build.sh
Running Jupyter-Book v0.11.2
Source Folder: /home/grg/src/msprime/docs
Config Path: /home/grg/src/msprime/docs/_config.yml
Output Path: /home/grg/src/msprime/docs/_build/html
Running Sphinx v3.3.1
[etoc] Changing master_doc to 'intro'
[etoc] Excluded 1 extra file(s) not in toc
loading pickled environment... done
myst v0.13.7: MdParserConfig(renderer='sphinx', commonmark_only=False, dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, update_mathjax=True, enable_extensions=['colon_fence', 'deflist'], disable_syntax=[], url_schemes=['mailto', 'http', 'https'], heading_anchors=None, html_meta=[], footnote_transition=True, substitutions=[], sub_delimiters=['{', '}'])
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 0 source files that are out of date
updating environment: executing outdated notebooks... done
0 added, 0 changed, 0 removed
looking for now-outdated files... none found
no targets are out of date.
build succeeded.

The HTML pages are in _build/html.

===============================================================================

Finished generating HTML for book.
Your book's HTML pages are here:
    _build/html/
You can look at your book by opening this file in a browser:
    _build/html/index.html
Or paste this line directly into your browser bar:
    file:///home/grg/src/msprime/docs/_build/html/index.html

===============================================================================

make: Leaving directory '/home/grg/src/msprime/docs'
```

--- a/requirements/development.txt
+++ b/requirements/development.txt
@@ -24,7 +24,7 @@ scipy
 # TODO we're pinning the version here because of problems on Travis.
 # versions weren't correctly being set for v1.12.1
 setuptools_scm == 1.11.1
-sphinx==3.3.1
+sphinx<4
 sphinx-argparse
 sphinx-issues
 jupyter-book

